vllm/model_executor/layers/fused_moe/fused_moe.py

```diff
@@ -455,51 +453,78 @@ def fused_experts(hidden_states: torch.Tensor,
+        tokens_in_chunk, _ = curr_hidden_states.shape
+
+        if tokens_in_chunk == 0:
+            break
+
+        if tokens_in_chunk < CHUNK_SIZE:
+            # will only happen in the last chunk
+            intermediate_cache1 = intermediate_cache1[:tokens_in_chunk]
+            intermediate_cache2 = intermediate_cache2[:tokens_in_chunk]
+            intermediate_cache3 = intermediate_cache3[:tokens_in_chunk]
+
+        curr_topk_ids = topk_ids[begin_chunk_idx:end_chunk_idx]
+        curr_topk_weights = topk_weights[begin_chunk_idx:end_chunk_idx]
+
+        sorted_token_ids, expert_ids, num_tokens_post_padded = moe_align_block_size(  # noqa: E501
+            curr_topk_ids, config['BLOCK_SIZE_M'], E)
+
+        invoke_fused_moe_kernel(curr_hidden_states,
+                                w1,
+                                intermediate_cache1,
+                                a1_scale,
+                                w1_scale,
+                                curr_topk_weights,
+                                curr_topk_ids,
+                                sorted_token_ids,
+                                expert_ids,
+                                num_tokens_post_padded,
+                                False,
+                                topk_ids.shape[1],
+                                config,
+                                compute_type=compute_type,
+                                use_fp8=use_fp8)
+
+        ops.silu_and_mul(intermediate_cache2, intermediate_cache1.view(-1, N))
+
+        invoke_fused_moe_kernel(intermediate_cache2,
+                                w2,
+                                intermediate_cache3,
+                                a2_scale,
+                                w2_scale,
+                                curr_topk_weights,
+                                curr_topk_ids,
+                                sorted_token_ids,
+                                expert_ids,
+                                num_tokens_post_padded,
+                                True,
+                                1,
+                                config,
+                                compute_type=compute_type,
+                                use_fp8=use_fp8)
+
+        if inplace:
+            torch.sum(intermediate_cache3.view(*intermediate_cache3.shape),
+                      dim=1,
+                      out=hidden_states[begin_chunk_idx:end_chunk_idx])
+        else:
+            states_acc.append(
+                torch.sum(intermediate_cache3.view(*intermediate_cache3.shape),
+                          dim=1))
     if inplace:
-        return torch.sum(intermediate_cache3.view(*intermediate_cache3.shape),
-                         dim=1,
-                         out=hidden_states)
-    return torch.sum(intermediate_cache3.view(*intermediate_cache3.shape),
-                     dim=1)
+        return hidden_states
+    else:
+        return torch.cat(states_acc, dim=0)
```

Comment:
I feel that we could pre-allocate a complete output tensor in advance and always perform in-place update to a slice, so that we could avoid chunk out allocation in each sum and the concat.

Reply:
Cool suggestion. This is only relevant for the is_place=False case right? In my impl I don't think there is any extra allocation being performed (in the in_place case)

Reply:
Yes this is only for `in_place=False` case.

Reply:
Great, yeah that's sort of what I just pushed in my commit. I used torch.zeros_like instead of empty, hope it's ok

Reply:
`empty_like` is better then `zeros_like` because `zeros` has to initialize tensor values which result in higher overheads.